File: Programmers/Lv. 3/섬 연결하기.py
def getParent(parent, x):
    if parent[x] == x:
        return x
    return getParent(parent, parent[x])
    
def unionParent(parent, a, b):
    a = getParent(parent, a)
    b = getParent(parent, b)
    if a < b:
        parent[b] = a
    else:
        parent[a] = b
        
def solution(n, costs):
    answer = 0
    
    costs.sort(key=lambda x : x[2])
    
    parent = [i for i in range(n)]
    
    lineCnt = 0
    for cost in costs:
        n1, n2, c = cost
        if getParent(parent, n1) != getParent(parent, n2):
            answer += c
            lineCnt += 1
            
            unionParent(parent, n1, n2)
            
        if lineCnt == n - 1:
            break
    
    return answer

File: Programmers/Lv. 3/섬 연결하기.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

vector<int> parent;

int getParent(int x)
{
    if (parent[x] == x) return x;
    return getParent(parent[x]);
}

void unionParent(int n1, int n2) 
{
    int a = getParent(n1);
    int b = getParent(n2);
    if (a > b) parent[a] = b;
    else parent[b] = a;
}

bool cmp(vector<int> a, vector<int> b)
{
    return a[2] < b[2];
}

int solution(int n, vector<vector<int>> costs) {
    int answer = 0;    
    
    for (int i = 0; i < n; i++)
    {
        parent.push_back(i);
    }
    
    sort(costs.begin(), costs.end(), cmp);
    
    for (auto cost : costs)
    {
        if (getParent(cost[0]) != getParent(cost[1]))
        {
            answer += cost[2];
            unionParent(cost[0], cost[1]);
        }
    }
    
    return answer;
}